[DR-3013] Add Terra Policy Service support

samanehsan opened the pull request:
https://broadworkbench.atlassian.net/browse/DR-3013

Add in a service class for the Terra Policy Service that has the ability to create and delete policy access objects. 

I was able to repurpose a lot of what the [workspace manager](https://github.com/DataBiosphere/terra-workspace-manager/blob/47512fdf63b3f168657f2993867c55aa0e86fa5f/service/src/main/java/bio/terra/workspace/service/policy/TpsApiDispatch.java) implemented. I ran into some trouble writing tests here, since it is not possible to mock the TPS api client that gets instantiated in the class so any ideas here would be helpful!

samanehsan committed:
Add policy service configuration

samanehsan committed:
Add Terra policy service

samanehsan committed:
Add status check for TPS

samanehsan committed:
Add policy service tests

samanehsan committed:
Fix typo and spotbugs

okotsopoulos commented on src/main/java/bio/terra/service/status/StatusService.java:
```diff
@@ -56,6 +61,7 @@ public RepositoryStatusModel getStatus() {
     statusModel.putSystemsItem(SAM, iamProviderInterface.samStatus().critical(true));
     statusModel.putSystemsItem(RBS, bufferService.status());
     statusModel.putSystemsItem(DUOS, duosService.status());
+    statusModel.putSystemsItem(TPS, policyService.status());
```
Nice call integrating this into our status check!

okotsopoulos commented on src/main/java/bio/terra/service/policy/PolicyService.java:
```diff
@@ -0,0 +1,139 @@
+  @Autowired
+  public PolicyService(PolicyServiceConfiguration policyServiceConfiguration) {
+    this.policyServiceConfiguration = policyServiceConfiguration;
+    logger.info("TPS base path: '{}'", policyServiceConfiguration.getBasePath());
```
I'm guessing this was added for debugging, would you mind removing or downgrading to a debug log?

okotsopoulos commented on src/main/java/bio/terra/service/policy/PolicyService.java:
```diff
@@ -0,0 +1,139 @@
+  private static final Logger logger = LoggerFactory.getLogger(PolicyService.class);
+  private final PolicyServiceConfiguration policyServiceConfiguration;
+
+  @Autowired
```
This @Autowired annotation is not needed in our version of Spring, since the class is annotated as a @Component.

okotsopoulos commented on src/main/java/bio/terra/service/policy/PolicyService.java:
```diff
@@ -0,0 +1,139 @@
+      throw new PolicyServiceAuthorizationException(
+          "Error reading or parsing credentials file", e.getCause());
+    }
+  }
```
To make it easier to unit test this service code, you could break out API construction (`getApiClient`, `getUnAuthApiClient`, `policyApi`) into its own service which can then be mocked in `PolicyServiceTest`.

I previously got this advice from @pshapiro4broad : https://github.com/DataBiosphere/jade-data-repo/pull/1314#discussion_r913827807

And implemented it here: https://github.com/DataBiosphere/jade-data-repo/pull/1314/commits/9790e6879180979db2f654dad4691255cf69279c

I've since followed this paradigm in other PRs and liked the ease of testing it offered.

okotsopoulos commented on src/main/java/bio/terra/service/policy/PolicyService.java:
```diff
@@ -0,0 +1,139 @@
+    }
+  }
+
+  @VisibleForTesting
+  public RuntimeException convertApiException(ApiException ex) {
+    if (ex.getCode() == HttpStatus.UNAUTHORIZED.value()) {
+      return new PolicyServiceAuthorizationException(
+          "Not authorized to access Terra Policy Service", ex.getCause());
+    } else if (ex.getCode() == HttpStatus.NOT_FOUND.value()) {
+      return new PolicyServiceNotFoundException("Policy service returns not found exception", ex);
+    } else if (ex.getCode() == HttpStatus.BAD_REQUEST.value()
+        && StringUtils.containsIgnoreCase(ex.getMessage(), "duplicate")) {
+      return new PolicyServiceDuplicateException(
+          "Policy service throws duplicate object exception", ex);
+    } else if (ex.getCode() == HttpStatus.CONFLICT.value()) {
+      return new PolicyConflictException("Policy service throws conflict exception", ex);
+    } else {
+      return new PolicyServiceAPIException(ex);
+    }
+  }
```
A few things:

- You could remove the `public` modifier and rely on the default visibility along with the `@VisibleForTesting` annotation.  That way your method is exposed for tests but not overexposed.
- This also looks like a method that should be static.
- I think you can streamline this method with a case switch expression.

```
  @VisibleForTesting
  static RuntimeException convertApiException(ApiException ex) {
    return switch (ex.getStatusCode()) {
      case HttpStatus.UNAUTHORIZED -> new PolicyServiceAuthorizationException(…);
      case HttpStatus.NOT_FOUND -> new PolicyServiceNotFoundException(…);
      …
      default -> new PolicyServiceAPIException(ex);
    };
  }
```

samanehsan replied:
I'm not sure if we can use case statements here since they need to call `.value()` to get the status code (e.g. `HttpStatus.UNAUTHORIZED.value()` instead of `HttpStatus.UNAUTHORIZED`), but IntelliJ complains that it needs to be a constant expression.

okotsopoulos replied:
Ah okay! The reason why I thought we could do this is that we have very similar handling in [`DuosClient`](https://github.com/DataBiosphere/jade-data-repo/blob/f6a8efcf99e5ae5ce85dc1507c7d91fe12ded6a1/src/main/java/bio/terra/service/duos/DuosClient.java#L119C1-L129), may be worth taking a look there.

okotsopoulos commented on src/main/java/bio/terra/service/policy/PolicyService.java:
```diff
@@ -0,0 +1,139 @@
+
+  public void createPao(
+      UUID resourceId, TpsObjectType resourceType, @Nullable TpsPolicyInputs policyInputs) {
+    policyServiceConfiguration.tpsEnabledCheck();
```
I'm curious about this "feature flag" -- what's the long-term intention?  Do we want environment-specific enforcement on whether or not these methods are called?  Ultimately would we want to remove this feature flag code when it's enabled everywhere?

samanehsan replied:
TPS is only available in dev right now, so this allows us to enable this per environment! We did something similar with RBS (although I think we never removed the enabled flag).

okotsopoulos replied:
Nice!

I’ve found that feature flag code can linger after you need it. What do you think about filing a preemptive ticket in the epic for its clean up, as well as updating the TPS hosts by environment? Awhile back we had a prod incident driven by me forgetting to update the ECM hosts by environment, and TDR prod mistakenly communicating with ECM dev.

samanehsan replied:
Good call! I created a ticket here: https://broadworkbench.atlassian.net/browse/DR-3029 and will update that with links to the code once everything is merged in.

okotsopoulos commented on src/main/java/bio/terra/service/policy/exception/PolicyServiceAPIException.java:
```diff
@@ -0,0 +1,47 @@
+import org.springframework.http.HttpStatus;
+
+/** Wrapper exception for non-200 responses from calls to Terra Policy Service. */
+public class PolicyServiceAPIException extends ErrorReportException {
```
Style nit: [Google Java style guide](https://google.github.io/styleguide/javaguide.html#s5.3-camel-case) advises camel casing of acronyms.  Under that style rule, this would be `PolicyServiceApiException`.

okotsopoulos commented on src/main/java/bio/terra/service/policy/PolicyService.java:
```diff
@@ -0,0 +1,139 @@
+      return new PolicyServiceAuthorizationException(
+          "Not authorized to access Terra Policy Service", ex.getCause());
+    } else if (ex.getCode() == HttpStatus.NOT_FOUND.value()) {
+      return new PolicyServiceNotFoundException("Policy service returns not found exception", ex);
```
Double-checking, do these exception wrappers communicate what we want to communicate to users in the event of an exception?

samanehsan replied:
Do you think something like `PolicyAccessObjectNotFoundException` is clearer? I borrowed this from the workspace manager and initially thought this meant the policy service itself could not be found (then I realized it's just your typical not found exception coming from the policy service).

okotsopoulos replied:
Ohhh sorry, I should have been more clear. I meant that as a comment on the exception messages! Ex. is "policy service returns not found exception" the message that we want to communicate to the user, or is there a more instructive message we can share with them?

okotsopoulos reviewed (commented):
I left some initial comments along with a suggestion on how to approach unit testing!

samanehsan committed:
Code review comments - cleanup/syntax changes

snf2ye commented on src/main/java/bio/terra/service/policy/exception/PolicyServiceAPIException.java:
It doesn't look like you're using all of these extensions of this exception. I tend to only add exception handling for known cases, but would be interested in your reasoning for adding all the possible cases. Do you have plans to use all of these cases? (i.e. I would only add `PolicyServiceAPIException(ApiException ex)`, `PolicyConflictException(String message, Throwable cause)`, and `getApiExceptionStatus` since I think these are the only ones in use )

samanehsan replied:
No, I can take them out! Good call!

samanehsan committed:
Create separate class for TPS get API methods

Separate these methods so they can be mocked in
the unit tests.

snf2ye commented on src/main/java/bio/terra/app/configuration/PolicyServiceConfiguration.java:
```diff
@@ -0,0 +1,52 @@
+  private boolean enabled;
+  private String basePath;
+
+  private static final List<String> BUFFER_SCOPES = List.of("openid", "email", "profile");
```
nit: Rename to `POLICY_SCOPES` or something to that affect.

snf2ye commented on build.gradle:
```diff
@@ -244,6 +244,8 @@ dependencies {
         }
     }
 
+    implementation group: "bio.terra", name: "terra-policy-client", version: "0.2.50-SNAPSHOT"
+    implementation group: "bio.terra", name: "terra-policy-service", version: "0.2.9-SNAPSHOT"
```
It's interesting that we need to add a reference to both the client and the service - Could we get away with just using the client?

samanehsan replied:
Oh you're totally right! Including the service was what caused the weird build errors.

samanehsan committed:
Add more policy servicee unit tests

samanehsan committed:
Fix naming conflict  with existing terra class

pshapiro4broad commented on src/main/java/bio/terra/service/policy/TerraPolicyService.java:
```diff
@@ -0,0 +1,114 @@
+
+  @VisibleForTesting
+  static RuntimeException convertApiException(ApiException ex) {
+    if (ex.getCode() == HttpStatus.UNAUTHORIZED.value()) {
```
If you use `org.apache.http.HttpStatus` instead you can convert this to an expression switch.
```
    return switch (e.getCode()) {
      case HttpStatus.SC_UNAUTHORIZED -> new PolicyServiceAuthorizationException(
          "Not authorized to access Terra Policy Service", e.getCause());
      case HttpStatus.SC_NOT_FOUND ->
          new PolicyServiceNotFoundException("Policy service returns not found exception", e);
      case HttpStatus.SC_BAD_REQUEST -> {
        if (StringUtils.containsIgnoreCase(e.getMessage(), "duplicate")) {
          yield new PolicyServiceDuplicateException(
              "Policy service throws duplicate object exception", e);
        } else {
          yield new PolicyServiceApiException(e);
        }
      }
      case HttpStatus.SC_CONFLICT ->
          new PolicyConflictException("Policy service throws conflict exception", e);
      default -> new PolicyServiceApiException(e);
    };
```

pshapiro4broad commented on src/main/java/bio/terra/service/policy/TerraPolicyService.java:
```diff
@@ -0,0 +1,114 @@
+  public void deletePao(UUID resourceId) {
+    policyServiceConfiguration.tpsEnabledCheck();
+    TpsApi tpsApi = policyApiService.getPolicyApi();
+    try {
+      try {
+        tpsApi.deletePao(resourceId);
+      } catch (ApiException e) {
+        throw convertApiException(e);
+      }
+    } catch (PolicyServiceNotFoundException e) {
+      // Not found should not cause a failure on delete.
+    }
```
I would avoid nesting `try` when possible. Since the code is catching an exception that's throwing, you can check the exception type before throwing instead. I think this should work
```suggestion
    try {
      tpsApi.deletePao(resourceId);
    } catch (ApiException e) {
      RuntimeException exception = convertApiException(e);
      if (PolicyServiceNotFoundException.class.isAssignableFrom(exception.getClass())) {
        // Not found should not cause a failure on delete.
        return;
      }
      throw exception;
    }
```

This is still not great IMO since it's code that's trying to re-create some state which other code generated and then discarded. One approach to fix this is to add an argument to `convertApiException()` that tells it if it should ignore the duplicate case. At that point you could make `convertApiException()` throw the exception itself, or have it return `null` or an `Optional` so the caller can tell when nothing needs to be thrown.

pshapiro4broad commented on src/test/java/bio/terra/service/policy/PolicyServiceTest.java:
```diff
@@ -0,0 +1,117 @@
+
+  @BeforeEach
+  public void setup() throws Exception {
+    lenient().when(policyServiceConfiguration.getEnabled()).thenReturn(true);
+    lenient().when(policyApiService.getPolicyApi()).thenReturn(tpsApi);
+    lenient().when(policyApiService.getUnauthPolicyApi()).thenReturn(tpsUnauthApi);
```
It'd be nice to avoid `lenient()` if possible. One approach is to add a method for each mocked API case and only set up the needed mocks before a test. This is what we did in the catalog code: https://github.com/DataBiosphere/terra-data-catalog/blob/main/common/src/test/java/bio/terra/catalog/rawls/RawlsServiceTest.java

samanehsan replied:
This looks much nicer!

samanehsan committed:
Revert "Fix naming conflict  with existing terra class"

This reverts commit 5982a60fc571622648502aa847551003e854b4af.

samanehsan committed:
Cleanup/address code review comments

- Remove policy service library (only need client)
- Remove unused exception code
- Improve policy service error messages
- Rename BUFFER_SCOPES -> POLICY_SCOPES in PolicyServiceConfiguration
- Use switch statement in convertApiException

samanehsan committed:
Remove lenient mocking

okotsopoulos commented on src/main/java/bio/terra/service/policy/PolicyApiService.java:
```diff
@@ -0,0 +1,44 @@
+public class PolicyApiService {
+  private final PolicyServiceConfiguration policyServiceConfiguration;
+
+  @Autowired
```
This `@Autowired` annotation can also be removed!

okotsopoulos reviewed (approved):
Looks great!  One small comment.

samanehsan committed:
Remove autowired annotation from PolicyApiService
